Redesigned long press menu

Stypox opened the pull request:
<!-- Hey there. Thank you so much for improving NewPipe, and filling out the details. Having roughly the same layout helps everyone considerably :)-->

#### What is it?
- \[ x \] Feature (user facing)
- \[ x \] Codebase improvement (dev facing)

#### Description of the changes in your PR

This PR replaces `InfoItemDialog` with a JetpackCompose-based bottom sheet menu. This new menu is supposed to replace all item long press menus across the app, i.e. streams, playlists, channels. The menu contains more details than before about the stream/playlist/channel that was clicked, laid out in a (imo) nice design.

Please provide feedback both on the code and on the UI, but please keep in mind the points below when suggesting an alternative layout. Note that the order of buttons is not yet finalized, and that at the moment the menu only works for streams.

<table>
  <tr>
    <th>Portrait</th>
    <th>Landscape/tablet</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/c6c1e7a4-a5f4-4d1a-abe6-6aef09563efd" width="300px" alt="Before"/></td>
    <td><img src="https://github.com/user-attachments/assets/8ccf2c89-b0e3-4e68-9b4f-178aedb32e6c" width="500px" alt="After"/></td>
  </tr>
</table>

#### Motivation and design goals

- A text-based menu is difficult to read and only allows for a handful of actions, while an icon-based menu is more compact and also easier to use. There still are some labels to explain what the actions do though.
- The menu puts the most frequent actions always in the same places, so muscle memory can help perform press the buttons faster.
- Users keep asking for new long press menu actions, therefore the long press actions system should be easily extendable and the set of actions should possibly even be editable by the user.
- We want to provide the same experience across the app (hence same menu for streams, playlists, channels), which also deduplicates a lot of code.
- A menu at the bottom of the screen is easier to use in one-handed mode.
- The menu adapts well to different screen sizes, e.g. by showing more/less buttons or adapting the header with item information.

<sup>Some of these points arose after some discussions in the NewPipe Matrix channel</sup>

#### TODO

- [ ] Also use the dialog for playlists and channels
- [ ] Perform asynchronous stuff with coroutines instead of RxJava3
- [ ] Implement "Start playing from here" actions
- [ ] Handle Kodi action
- [ ] Decide whether to allow the user to customize the menu (there is already a mini settings button in the top right of the menu for that purpose, but now it doesn't do anything)
- [ ] Various other TODOs spread out in code
- [ ] Understand which issues are solved by this PR
- [ ] Add documentation
- [ ] Add UI tests (?)

<!--
#### Fixes the following issue(s)
- Fixes # -->

#### APK testing
<!-- Use a new, meaningfully named branch. The name is used as a suffix for the app ID to allow installing and testing multiple versions of NewPipe, e.g. "commentfix", if your PR implements a bugfix for comments. (No names like "patch-0" and "feature-1".)  -->
<!-- Remove the following line if you directly link the APK created by the CI pipeline. Directly linking is preferred if you need to let users test.-->
The APK can be found by going to the "Checks" tab below the title. On the left pane, click on "CI", scroll down to "artifacts" and click "app" to download the zip file which contains the debug APK of this PR. You can find more info and a video demonstration [on this wiki page](https://github.com/TeamNewPipe/NewPipe/wiki/Download-APK-for-PR).

#### Due diligence
- \[ x \] I read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md).

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/ui/components/menu/LongPressMenu.kt:
```diff
@@ -0,0 +1,586 @@
+                    }
+                }
+
+            Column {
```
[`FlowRow`](https://developer.android.com/develop/ui/compose/layouts/flow) could be used here instead.

Stypox replied:
Mmmh, that's what I was using in the beginning, but I don't think you can have a FlowRow with a maximum amount of items per row, but also have an item that weighs more than the others (i.e. the header). Also, I wanted the buttons to expand as much as possible, to not waste any space.

Isira-Seneviratne replied:
The linked documentation mentions a `maxItemsInEachRow` attribute.

Stypox replied:
I know, but I need the header to weigh as if it were 4 items

Isira-Seneviratne replied:
I was able to make the needed changes using FlowRow.

Stypox replied:
Try now

Isira-Seneviratne replied:
Still not working

Isira-Seneviratne replied:
Here's the patch:

[Handle_LongPressMenu_layout_using_FlowRow.patch](https://github.com/user-attachments/files/18802258/Handle_LongPressMenu_layout_using_FlowRow.patch)

Stypox replied:
Mmmh, I should have given you write access to the NewPipe repository (or maybe you had it already), which should be enough according to what GitHub says:

> [...] users with write access to TeamNewPipe/NewPipe can add new commits to your long-press-menu branch.

Anyways, thank you for the patch! I haven't tested it yet (will do tomorrow) but from a brief look it seems like the fact that the header is always 66% of the width means the buttons in the first row won't be aligned like the ones below. Also, the buttons should fill as much horizontal space as possible, so misclicks can still be caught, and there might also be more space for the button label text. One final (minor) thing is that I wanted the layout to scale well just based on the available space, not specifically based on whether the device is landscape or portrait, so the header sizing in your patch does not fulfill this requirement.

Isira-Seneviratne replied:
> One final (minor) thing is that I wanted the layout to scale well just based on the available space, not specifically based on whether the device is landscape or portrait, so the header sizing in your patch does not fulfill this requirement.

That can be implemented using `currentWindowAdaptiveInfo().windowSizeClass`, like in the `determineItemViewMode()` method.

Stypox replied:
Sure, but what's bad about using BoxWithConstraints and measuring the actual available space?

Isira-Seneviratne replied:
Using it results in the IDE checking if the constraints are actually used.

Stypox replied:
You are right that it is indeed much more complex, so we might want to change the requirements. In the meantime I will add more documentation on the requirements (so people reading the code know why it was made like that).

Stypox commented:
@NaN-Sz thanks for the feedback! We already thought of the More button but the bottom sheet dialog already has kind of a way to show more buttons when there are too many (i.e. the menu starts out partially expanded, and you can scroll up to see the other buttons). What do you think about that?

Also, do you think the buttons are too small in my screenshots? Your mockup has much bigger ones, that's why I'm asking. I find the buttons in my screenshots easy to press correctly but it's arguable. Maybe we could put a slider in the "long-press menu's settings menu" to pick a size for buttons. This would be quite a niche setting, but also easy to understand what it does (and we would put it only in the "long-press menu's settings menu", not in the app's settings).

shivang1809 commented:
> or maybe like this?
> 
> <img alt="" width="200px" src="https://private-user-images.githubusercontent.com/120607988/418305816-e7361929-622d-410b-b510-f39aaa6c9326.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI2MjgwODksIm5iZiI6MTc0MjYyNzc4OSwicGF0aCI6Ii8xMjA2MDc5ODgvNDE4MzA1ODE2LWU3MzYxOTI5LTYyMmQtNDEwYi1iNTEwLWYzOWFhYTZjOTMyNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMyMlQwNzE2MjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMTEwMzgwNGY5NzUwNTA1MTcxNzdhMDc4NmNkMWZjNDg5MDA5ZGI2ZDJjOGVhNDY0NWI4ZjA4ODZhNWU3NzBkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.IqB5h3jETsOU2z4eX5_-Kd11xwgOHQxWeBwCCBc0mvY">

Would it be a good idea to make it like only 3 or 4 most used options come on long press (delete, add to playlist, share, background ,etc). And then could be swiped up to show this all like it is. This will reduce vertical space. 

A cherry on cake will be if there's a customize option to chose which options should come on long press and all other in swipe up menu.

MD77MD commented:
the most important option is missing : Open Channel

Stypox commented:
@MD77MD no, it is still there, it was moved under the video name to save some space (it should be noticeable since it's colored in blue)

Stypox commented:
@MD77MD would an icon like this make it clearer?

<img width="471" height="106" alt="image" src="https://github.com/user-attachments/assets/5bc09869-d399-46c6-a5bc-e178cbf9d03b" />

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/ui/components/menu/LongPressAction.kt:
I believe it'd be a good idea to introduce separate Stream and Playlist data classes, like how I've done here: https://github.com/TeamNewPipe/NewPipe/pull/11947

That would remove the need to define multiple factory methods.

Stypox commented:
I added three more actions and I need your feedback on them:
- They allow playing in background/popup/main-player a list of items (e.g. a playlist) starting from the currently selected item.
- This feature is already present in some menus of the non-refactored NewPipe, but not in all of them, and when it exists it replaces the corresponding action to play just the single item.
- Feedback is needed on the icons: I made them up by combining the base icon (headset, picture in picture, and play) with a "badge" icon from Material Icons, let me know if you can think of a better "badge" icon from the ones available [here](https://fonts.google.com/icons?icon.query=).
- Now you can play-from-here in the feed and in channels, which is currently not possible in non-refactored NewPipe! 

<img width="1080" height="1920" alt="image" src="https://github.com/user-attachments/assets/096a852a-bdf7-4a02-b05d-e0e835513487" />

KraXen72 commented:
i would suggest using the arrow-forward icon as the badge, and intersect it with the full-sized icon, same as enqueue next does. 

also, the labels overflowing doesen't look great, i'd keep it in one line if possible

otherwise, can't wait for this new bottom menu! looks awesome!

Stypox commented:
Here are 4 alternatives based on suggestions on Matrix and on @KraXen72's suggestions.
- The first and the fourth one look good imo, however the arrow-forward icon might not clearly indicate the intention to "play from here" (?)
- The second and the third look too crowded

<img width="356" height="115" alt="image" src="https://github.com/user-attachments/assets/806ed40e-2935-48c1-a1c3-51ceeb3c65af" />

<img width="356" height="115" alt="image" src="https://github.com/user-attachments/assets/8f4723e6-3f31-428b-94b8-9f1878afdd3f" />

<img width="356" height="115" alt="image" src="https://github.com/user-attachments/assets/e74c22b4-6e05-40d8-8ed0-5054776cb06b" />

<img width="356" height="115" alt="image" src="https://github.com/user-attachments/assets/10a304f3-5758-478b-9d57-1687ce7adf40" />


> also, the labels overflowing doesen't look great, i'd keep it in one line if possible

Since "Background from here" is a quite long text I don't think there is a way to keep it to one line. But I moved the three icons a bit higher, so there is more space above the text, and I think it looks better now.

Isira-Seneviratne commented on app/src/main/java/org/schabi/newpipe/ui/components/menu/LongPressAction.kt:
```diff
@@ -0,0 +1,320 @@
+                            listOf(Type.Unsubscribe.buildAction { onUnsubscribe.run() })
+                        }
+                        ?: listOf()
+                    )
```
```suggestion
                listOfNotNull(onUnsubscribe?.let { listOf(Type.Unsubscribe.buildAction { it.run() }) })
```

Stypox commented:
I mostly implemented the long-press menu actions editor, where you can enable and disable actions, and even enable and disable the header. The order is still not persisted anywhere, so next thing I'll do is store the data in settings. There are also a few TODOs left, e.g. making the UI usable with DPAD buttons (e.g. Android TV). I borrowed some code from [FlorisBoard](https://github.com/florisboard/florisboard/blob/059d2fd4bfa9f30f634552f4457c4541ba44a238/app/src/main/kotlin/dev/patrickgold/florisboard/ime/smartbar/quickaction/QuickActionsEditorPanel.kt), so thanks! :-) 
![Screen_recording_20251022_011336](https://github.com/user-attachments/assets/b06aa57b-2a06-43c3-8179-5944b8bad9e7)

Stypox commented:
I'm designing the icon for the "Play in the background all items shuffled" that would appear when long-pressing playlists or channels. Which one of these designs should I choose? They all vary only by the thickness of the arrows in the bottom-right icon. 

<img width="400px" alt="image" src="https://github.com/user-attachments/assets/65bde72e-927e-4042-b2c2-e285038890c5" />


<sub>Note for future me: the corresponding thinkness increments are 0, 0.08, 0.12, 0.15, 0.18, 0.21, 0.25, 0.3, 0.34.</sub>

Stypox commented:
Ok, this PR is not fully ready for review and user feedback :-D

- The editor is now fully functional, it works with touch and DPAD, and it saves the button arrangement to settings
- I added the "Background/Popup/Play shuffled" options (I ended up choosing thickness 3), which fetches a few pages of the queue before shuffling it, to solve #875
- I wrote quite a few UI tests and normal tests (~1500 lines of code)
- I cleaned up the code and added javadocs everywhere, so hopefully it should be rather easy to review despite the size
- I updated the PR description with screenshots
- Sorry for the 87 commits :sweat_smile:, though I believe they shouldn't be squashed as they contain useful information

theimpulson commented on app/src/main/java/org/schabi/newpipe/ui/components/common/Tooltip.kt:
```diff
@@ -0,0 +1,84 @@
+ */
+@Composable
+fun TooltipIconButton(
+    onClick: () -> Unit,
```
onClick can be made last parameter to allow better kotlin syntax when calling it

TobiGr commented on app/src/main/java/org/schabi/newpipe/player/playqueue/ChannelTabPlayQueue.java:
```diff
@@ -15,22 +19,39 @@
 
     public ChannelTabPlayQueue(final int serviceId,
                                final ListLinkHandler linkHandler) {
         this(serviceId, linkHandler, null, Collections.emptyList(), 0);
```
I'd also rename the ListLinkHandler to `tabHandler` to be consistent.
